### PR TITLE
[IMP] *: payment screen new design

### PR DESCRIPTION
**Pull request by taal-odoo**
*: point_of_sale, pos_online_payment, pos_restaurant

Before this commit:
---
[
before
](url)

After this commit:
---
Revised graphical interface for payment screen and updated test tours to match

after

task-3888884

https://github.com/odoo/enterprise/pull/62250

### CODE REVIEW

**Comment by andrewdhastings** [comment]
Wouldn't it make more sense to put the key pad on the right hand side as most people are right handed? Always bugged me about odoo pos layout. I know it is a small thing, but always having to cross your body with your hand when using a touch screen always feels odd.

**Comment by vlst-odoo** [comment]
> Wouldn't it make more sense to put the key pad on the right hand side as most people are right handed? Always bugged me about odoo pos layout. I know it is a small thing, but always having to cross your body with your hand when using a touch screen always feels odd.

:exploding_head: 
good point

File: addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/** @odoo-module */
+
+import { NumberPopup } from "@point_of_sale/app/utils/input_popups/number_popup";
+import { patch } from "@web/core/utils/patch";
+import { usePos } from "@point_of_sale/app/store/pos_hook";
+
+patch(NumberPopup.prototype, {
+    setup() {
+        super.setup(...arguments);
+        this.pos = usePos();
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.js:10]: `NumberPopup` should not depend on pos.
pass whatever you need through props

File: addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates id="template" xml:space="preserve">
+    <t t-name="pos_restaurant.NumberPopup" t-inherit="point_of_sale.NumberPopup" t-inherit-mode="extension">
+        <xpath expr="//button[hasclass('o-default-button')]" position="after">
+            <div class="d-flex flex-grow-1" />
+            <t t-esc="amountPerGuest"/> / Guest
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.xml:6]:
```suggestion
            <div class="ms-auto">
                <t t-esc="amountPerGuest" /> / Guest
            </div> 
```

File: addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js [hunk 3]
```diff
@@ -16,6 +16,11 @@ patch(PaymentScreen.prototype, {
         }
         return super.nextScreen;
     },
+    get amountPerGuest() {
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js:19]: this same function appears in multiple places. Find a way to not repeat it. Maybe put it on the order model

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_status/payment_status.xml [hunk 4]
```diff
@@ -2,42 +2,32 @@
...
             </div>
-
-            <t t-else="">
+        </t>
+        <t t-else="">
+            <section class="paymentlines-container p-3 border-bottom">
                 <div class="payment-status-container d-flex flex-column-reverse flex-lg-row justify-content-between fs-2">
-                    <div>
-                        <div class="payment-status-remaining">
+                    <t t-if="this.props.order.get_due() >= 0 and props.order.get_change() === 0">
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_status/payment_status.xml:13]: you might want to use `gte` here.
[reference](https://github.com/odoo/owl/blob/master/doc/reference/templates.md#:~:text=it%20can%20use%20a%20few%20special%20operators%20to%20avoid%20using%20symbols%20such%20as%20%3C%2C%20%3E%2C%20%26%20or%20%7C.%20This%20is%20useful%20to%20make%20sure%20that%20we%20still%20write%20valid%20XML.)

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss [hunk 5]
```diff
@@ -1,5 +1,34 @@
...
     vertical-align: middle;
 }
+
+.paymentmethods-small {
+  max-height: 186px;
+}
+
+.left-content .numpad button {
+    font-size: 1.1rem;
+}
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss:17]: can you have a way to not need this custom css? Preferably with bootstrap

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

**Comment by mobt-odoo** [comment]
> > Wouldn't it make more sense to put the key pad on the right hand side as most people are right handed? Always bugged me about odoo pos layout. I know it is a small thing, but always having to cross your body with your hand when using a touch screen always feels odd.
> 
> 🤯 good point @mobt-odoo what do you think ?


> > Wouldn't it make more sense to put the key pad on the right hand side as most people are right handed? Always bugged me about odoo pos layout. I know it is a small thing, but always having to cross your body with your hand when using a touch screen always feels odd.
> 
> 🤯 good point @mobt-odoo what do you think ?

Thanks for the relevant feedback. Nonetheless, the human eye tends to find it confortable to look on the right. Therefore, we focus on putting the most important infos on the right. At this point the most important action to do in PoS, is to be able to see the all the product you may potentially select.

File: addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js (hunk 3 above)
Comment (taal-odoo) [reply addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js:19]: The implementation of the `amountPerGuest` function in the numpad.js depends on the state of the buffer to make it responsive to input, which differs from the one in the payment_screen.

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss (hunk 5 above)
Comment (taal-odoo) [reply addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss]: There doesn't appear to be a Bootstrap class for the desired font size. Additionally, this font size adjustment needs to be applied exclusively to the numpad buttons on the payment screen, without affecting other screens.

File: addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js (hunk 3 above)
Comment (vlst-odoo) [reply addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.js:19]: That should not be a problem. Look [here](https://github.com/odoo/odoo/blob/master/addons/point_of_sale/static/src/app/utils/input_popups/number_popup.js#L15) as an example

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss (hunk 5 above)
Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.scss]: > There doesn't appear to be a Bootstrap class for the desired font size.

I am sure that among all the bootstrap utilities you will find what you need.

> Additionally, this font size adjustment needs to be applied exclusively to the numpad buttons on the payment screen, without affecting other screens

You can give a custom class to the buttons. [Reference](https://github.com/odoo/odoo/blob/master/addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js#L13)

**Comment by taal-odoo** [comment]
@vlst-odoo  Thanks for the suggestions, changes have been made.

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js [hunk 6]
```diff
@@ -19,33 +19,60 @@ export const buttonsType = {
         String,
     ],
 };
-export function getButtons(env, rightColumn) {
+export function getButtons(env, rightColumn, options = {}) {
```
[the remark this answers is not in the document]
Comment (taal-odoo) [reply addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js]: I have made the change feel free to review them

File: addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js [hunk 7]
```diff
@@ -33,10 +33,10 @@ export function getButtons(env, rightColumn) {
         { value: "8" },
         { value: "9" },
         ...(rightColumn ? [rightColumn[2]] : []),
-        ...(rightColumn ? [rightColumn[3]] : []),
-        { value: "0" },
         { value: env.services.localization.decimalPoint },
+        { value: "0" },
         { value: "Backspace", text: "⌫" },
+        ...(rightColumn ? [rightColumn[3]] : []),
     ];
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js:40]: no need for this change

File: addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js [hunk 8]
```diff
@@ -27,6 +27,13 @@ patch(ControlButtons.prototype, {
         this.dialog.add(NumberPopup, {
             startingValue: this.currentOrder?.getCustomerCount() || 0,
             title: _t("Guests?"),
+            feedback: (buffer) => {
+                if (buffer && buffer.length > 0) {
+                    const guestCount = parseInt(buffer, 10) || 0;
+                    this.currentOrder.setCustomerCount(guestCount);
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js:33]: you should never mutate in the feedback.
Also you probably want the return to be something like ``value ?  `Amount per customer: ${value}` : '' ``

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js (hunk 8 above)
Comment (taal-odoo) [reply addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js]: Sure, my bad for the earlier oversight on mutating data in the feedback function, which has now been corrected.

Regarding the return value, the `this.env.utils.formatCurrency` function manages the currency symbol. Changes in XML were made to align with the specified positioning and supporting text requirements.

File: addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js (hunk 7 above)
Comment (mobt-odoo) [reply addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js:40]: Thanks

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.xml [hunk 9]
```diff
@@ -64,59 +64,104 @@
...
+            </div>
+        </t>
+    </t>
+
+    <t t-name="point_of_sale.PaymentScreenBottom">
+        <div t-attf-class="d-flex {{ui.isSmall ? 'switchpane' : 'flex-row'}}">
+            <t t-call="point_of_sale.PaymentScreenValidate" />
+            <t t-call="point_of_sale.PaymentScreenBack" />
         </div>
     </t>
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.xml:90]: if this is only used once there's no need to have a separate template for it. You can inline it

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.xml [hunk 10]
```diff
@@ -64,59 +64,104 @@
...
+                                <div>
+                                    <img class="payment-method-icon" t-att-src="paymentMethodImage(paymentMethod.id)" />
+                                </div>
+                                <span class="payment-name" t-esc="paymentMethod.name" />
+                            </div>
+                        </div>
+                    </t>
+                </div>
+            </div>
+        </t>
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_screen.xml:165]: can you find a way to unify these 2 for loops into one?

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_status/payment_status.xml [hunk 11]
```diff
@@ -2,42 +2,32 @@
...
-        <section class="paymentlines-container bg-view p-3 mb-1 border-bottom">
-            <div t-if="props.order.get_paymentlines().length === 0" class="paymentlines-empty">
-                <div class="total text-center py-2 text-success">
-                    <t t-esc="totalDueText" />
-                </div>
-                <div class="message text-center">
-            Please select a payment method.
-                </div>
+        <t t-if="props.order.get_paymentlines().length == 0">
+            <div class="message text-center py-4 fs-4">
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_status/payment_status.xml:6]:
```suggestion
            <div t-if="props.order.get_paymentlines().length == 0" class="message text-center py-4 fs-4">
```

Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/screens/payment_screen/payment_status/payment_status.xml:6]: prefer having fewer tags when possible

File: addons/pos_restaurant/static/src/overrides/models/models.js [hunk 12]
```diff
@@ -42,6 +42,13 @@ patch(Order.prototype, {
     setCustomerCount(count) {
         this.customerCount = Math.max(count, 0);
     },
+    amountPerGuest() {
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/models/models.js:45]:
```suggestion
    amountPerGuest(numCustomers = this.customerCount) {
```
this way the code in the numberPopup feedback will be much simpler

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js (hunk 7 above)
Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/generic_components/numpad/numpad.js:40]: this change still has to be reverted. The numpad changed in master after the printscreen was taken for the spec. That's why they differ, not intentionally

File: addons/point_of_sale/static/src/app/screens/partner_list/partner_line/partner_line.xml
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/partner_list/partner_line/partner_line.xml:1]: given that this component is not changed in the spec, why is this change necessary ?

File: addons/point_of_sale/static/src/app/screens/partner_list/partner_list.xml
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/screens/partner_list/partner_list.xml:1]: same question as for partner line

File: addons/pos_restaurant/static/src/overrides/models/models.js [hunk 13]
```diff
@@ -42,6 +42,13 @@ patch(Order.prototype, {
     setCustomerCount(count) {
         this.customerCount = Math.max(count, 0);
     },
+    amountPerGuest(numCustomers = this.customerCount) {
+        if (numCustomers === 0) {
+            return this.env.utils.formatCurrency(0);
+        }
+        const totalDue = this.get_total_with_tax() + this.get_rounding_applied();
+        return this.env.utils.formatCurrency(totalDue / numCustomers);
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/models/models.js:50]: in general your methods should return the numeric value and the you would format in the template. This way you can use that value in other computations too

File: addons/pos_restaurant/static/src/overrides/models/models.js [hunk 14]
```diff
@@ -42,6 +42,13 @@ patch(Order.prototype, {
     setCustomerCount(count) {
         this.customerCount = Math.max(count, 0);
     },
+    amountPerGuest(numCustomers = this.customerCount) {
+        if (numCustomers === 0) {
+            return this.env.utils.formatCurrency(0);
+        }
+        const totalDue = this.get_total_with_tax() + this.get_rounding_applied();
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/models/models.js:49]: you can extract this into a separate method in the pos order model and get rid of the getter in payment screen

File: addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.xml
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/components/input_popups/number_popup.xml:1]: remove this file. Just use the normal feedback mechanism ( same as in pos sale )
This will mean that the amount will be shown just underneath the input box, instead of in the footer, but that is fine. We don't need 2 ways to show the feedback.

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

**Comment by taal-odoo** [comment]
Pushed with required changes

File: addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js [hunk 15]
```diff
@@ -25,6 +25,12 @@ patch(ControlButtons.prototype, {
         this.dialog.add(NumberPopup, {
             startingValue: this.currentOrder?.getCustomerCount() || 0,
             title: _t("Guests?"),
+            feedback: (buffer) => {
+                const value = this.env.utils.formatCurrency(
+                    this.currentOrder.amountPerGuest(parseInt(buffer, 10) || 0)
+                );
+                return value ? `${value} / Guest` : "";
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js:32]: `Guest` should be translatable

File: addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.xml [hunk 16]
```diff
@@ -13,14 +13,22 @@
...
                 <span class="back_text">Keep Open</span>
             </t>
             <t t-else="">$0</t>
         </xpath>
     </t>
-
+    <t t-name="pos_restaurant.PaymentScreenDue" t-inherit="point_of_sale.PaymentScreenDue" t-inherit-mode="extension">
+        <xpath expr="//section[hasclass('paymentlines-container')]" position="inside">
+            <t t-if="currentOrder.getCustomerCount()">
+                <div class="message text-center fs-4">
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.xml:27]:
```suggestion
            <div t-if="currentOrder.getCustomerCount()" class="message text-center fs-4">
```

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js (hunk 15 above)
Comment (taal-odoo) [reply addons/pos_restaurant/static/src/app/control_buttons/control_buttons.js]: Done.

File: addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.xml (hunk 16 above)
Comment (taal-odoo) [reply addons/pos_restaurant/static/src/overrides/components/payment_screen/payment_screen.xml]: I have made the correction.

**Review by vlst-odoo — APPROVED** [review]
robodoo r+

**Comment by vlst-odoo** [comment]
robodoo r-

**Comment by vlst-odoo** [comment]
robodoo r+

**Comment by taal-odoo** [comment]
@robodoo r+

**Comment by vlst-odoo** [comment]
robodoo r+
